File: 0525-contiguous-array/0525-contiguous-array.java
class Solution {
    public int findMaxLength(int[] arr) {
        int n=arr.length;
        HashMap<Integer,Integer>mp=new HashMap<>();
        mp.put(0,-1);
        int s=0;int maxlen=0;
        for(int i=0;i<n;i++){
            if(arr[i]==1){
                s++;
            }else{
                s--;
            }
            if(mp.containsKey(s)){
                maxlen=Math.max(maxlen,i-mp.get(s));
            }else{
                mp.put(s,i);
            }
            
        }
        return maxlen;
        
    }
}

File: 0525-contiguous-array/0525-contiguous-array.cpp
class Solution {
public:
    int findMaxLength(vector<int>& arr) {
        int n=arr.size();
        int s=0;
        int maxlen=0;
        unordered_map<int, int> mp;
        mp[0]=-1;
        for(int i=0;i<n;i++){
            s+=(arr[i]==0)?1:-1;
            if(mp.find(s)!=mp.end()){
                maxlen=max(maxlen,(i-mp.at(s)));
            }
            else{
                mp[s]=i;
            }

        }
        return maxlen;
    }
};